lib/pengine/pe_actions.c

```diff
@@ -446,66 +446,87 @@ validate_on_fail(const pe_resource_t *rsc, const char *action_name,
+         */
+        for (xmlNode *operation = first_named_child(rsc->ops_xml, XML_ATTR_OP);
+             (operation != NULL); operation = crm_next_same_xml(operation)) {
             bool enabled = false;
+            const char *promote_on_fail = NULL;
 
-            if (!pcmk__str_eq((const char *)operation->name, "op", pcmk__str_none)) {
+            /* We only care about explicit on-fail (if promote uses default, so
+             * can demote)
+             */
+            promote_on_fail = crm_element_value(operation, XML_OP_ATTR_ON_FAIL);
+            if (promote_on_fail == NULL) {
                 continue;
             }
+
+            // We only care about recurring monitors for the promoted role
             name = crm_element_value(operation, "name");
             role = crm_element_value(operation, "role");
-            on_fail = crm_element_value(operation, XML_OP_ATTR_ON_FAIL);
-            interval_spec = crm_element_value(operation, XML_LRM_ATTR_INTERVAL);
-            if (!on_fail) {
-                continue;
-            } else if (pcmk__xe_get_bool_attr(operation, "enabled", &enabled) == pcmk_rc_ok && !enabled) {
+            if (!pcmk__str_eq(name, PCMK_ACTION_MONITOR, pcmk__str_none)
+                || !pcmk__strcase_any_of(role, PCMK__ROLE_PROMOTED,
+                                         PCMK__ROLE_PROMOTED_LEGACY, NULL)) {
                 continue;
-            } else if (!pcmk__str_eq(name, PCMK_ACTION_MONITOR, pcmk__str_casei)
-                       || !pcmk__strcase_any_of(role, PCMK__ROLE_PROMOTED,
-                                                PCMK__ROLE_PROMOTED_LEGACY,
-                                                NULL)) {
+            }
+            interval_spec = crm_element_value(operation, XML_LRM_ATTR_INTERVAL);
+            if (crm_parse_interval_spec(interval_spec) == 0) {
                 continue;
-            } else if (crm_parse_interval_spec(interval_spec) == 0) {
+            }
+
+            // We only care about enabled monitors
+            if ((pcmk__xe_get_bool_attr(operation, "enabled",
+                                        &enabled) == pcmk_rc_ok) && !enabled) {
                 continue;
-            } else if (pcmk__str_eq(on_fail, "demote", pcmk__str_casei)) {
+            }
+
+            // Demote actions can't default to on-fail="demote"
+            if (pcmk__str_eq(promote_on_fail, "demote", pcmk__str_casei)) {
                 continue;
             }
 
-            value = on_fail;
+            // Use value from first applicable promote action found
+            return on_fail;
         }
-    } else if (pcmk__str_eq(action_name, PCMK_ACTION_LRM_DELETE,
-                            pcmk__str_casei)) {
-        value = "ignore";
+        return NULL;
+    }
 
-    } else if (pcmk__str_eq(value, "demote", pcmk__str_casei)) {
+    if (pcmk__str_eq(action_name, PCMK_ACTION_LRM_DELETE, pcmk__str_none)
+        && !pcmk__str_eq(value, "ignore", pcmk__str_none)) {
```

Comment:
We're still doing case-insensitive for on-fail values, right?

Reply:
Good catch